_source/_posts/2022-01-13-spring-cloud-contract.md

```diff
@@ -0,0 +1,893 @@
+
+A brief aside on "mocks" and "stubs". "Mock" gets used in ways online that is somewhat generic, meaning any fake object used for testing, and this can get confusing when differentiating "mocks" from "stubs". However, specifically, a "mock" is an object that tests for behavior by registering expected method calls before a test run while a "stub" is a testable version of the object with callable methods that return pre-set values. Thus, a mock checks to see if the object being tested makes an expected sequence of calls to the object being mocked, and throws an error if the behavior deviates (that is, makes any unexpected calls), while a stub does not do any testing itself, per se, but instead will return canned responses to pre-determined methods to allow tests to run (of course, it can throw an error if unexpected calls are made as well, so stubs do constrain behavior in that sense as well). 
+
+In this tutorial, you're going to create a simple Spring Boot **producer** that serves hats and a Spring Boot **consumer** that presents hats to the user (really hat specification object instances). The consumer will retrieve the hats from the producer before presenting them. You will use Spring Cloud Contract to write a contract that describes what the producer service should return based on certain resource queries. Spring Cloud Contract will automatically generate integration tests from the producer contract, which will verify that the producer is behaving according to the specification of the contract, and will also generate stubs that can be imported into other projects to assist in integration testing. Next, on the consumer side, you will import these stubs into the consumer's integration tests as a mocked object that can be used in testing the consumer. You will also see how Okta can be used as an OAuth 2.0 and OpenID Connect (OIDC) provider to add security to the Spring Boot project.
+
+**Prerequisites:**
+
+Before you get started, you'll need to make sure you have a few tools installed.
+
+- **Okta CLI**: The Okta CLI is an easy way to create projects that use Okta for security. Follow the installation instructions on the [Okta CLI project website](https://cli.okta.com/manual/#installation). You should log in to your existing account or register for a new account using the CLI before continuing with the tutorial.
+
+- **Java 11**: This project uses Java 11. OpenJDK 11 will work just as well.  Instructions are found on the [OpenJDK website](https://openjdk.java.net/install/). OpenJDK can also be installed using [Homebrew](https://brew.sh/). Alternatively, [SDKMAN](https://sdkman.io/) is another excellent option for installing and managing Java versions.
+
+- **HTTPie**: This is a powerful command-line HTTP request utility you'll use to test both Spring Boot resource servers. Install it according to [the docs on their site](https://httpie.org/doc#installation).
+
+## Bootstrap the Producer Project using Spring Initialzr
```

Comment:
```suggestion
## Bootstrap the producer project using Spring Initialzr
```
switched to sentence case for all article headers to align with broader Okta Dev style guide. Only article titles use title case.